File: functions/api/utils/notificationService.js
const { admin, db } = require('../config/firebase');

/**
 * Reusable Helper to Send Push Notifications
 */
async function sendNotification(recipientId, title, body, data = {}, category = 'system') {
    try {
        if (!recipientId) return;

        const userDoc = await db.collection('users').doc(recipientId).get();
        if (!userDoc.exists) {
            console.log(`Notification skipped: User ${recipientId} not found`);
            return;
        }

        const userData = userDoc.data();
        const fcmToken = userData.fcmToken;

        // 1. Audit Log (Always save to Firestore)
        await db.collection('notifications').add({
            recipientId,
            title,
            body,
            data,
            category,
            status: fcmToken ? 'sent' : 'pending_token',
            sentAt: admin.firestore.FieldValue.serverTimestamp(),
            read: false
        });

        // 2. Send Push if Token Exists
        if (fcmToken) {
            const message = {
                token: fcmToken,
                notification: { title, body },
                data: {
                    click_action: 'FLUTTER_NOTIFICATION_CLICK',
                    category,
                    ...data,
                },
                android: {
                    priority: 'high',
                    notification: { channelId: 'high_importance_channel' }
                },
                apns: {
                    payload: { aps: { sound: 'default', contentAvailable: true } }
                }
            };
            await admin.messaging().send(message);
            console.log(`Notification sent to ${recipientId}: ${title}`);
        }
    } catch (error) {
        console.error("Send Notification Helper Error:", error);
    }
}

module.exports = { sendNotification };

File: lib/logic/services/notification_service.dart
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/foundation.dart';

class NotificationService {
  static final NotificationService _instance = NotificationService._internal();
  factory NotificationService() => _instance;
  NotificationService._internal();

  static const String _baseUrl = 'https://functions-kappa-gold.vercel.app';

  final FirebaseMessaging _fcm = FirebaseMessaging.instance;
  final FlutterLocalNotificationsPlugin _localNotifications =
      FlutterLocalNotificationsPlugin();

  Future<void> initialize() async {
    // 1. Request Permissions
    NotificationSettings settings = await _fcm.requestPermission(
      alert: true,
      badge: true,
      sound: true,
    );

    if (settings.authorizationStatus == AuthorizationStatus.authorized) {
      debugPrint('User granted notification permissions');
    }

    // 2. Setup Local Notifications (for Foreground)
    const AndroidInitializationSettings androidSettings =
        AndroidInitializationSettings('@mipmap/ic_launcher');
    const DarwinInitializationSettings iosSettings =
        DarwinInitializationSettings();

    const InitializationSettings initSettings =
        InitializationSettings(android: androidSettings, iOS: iosSettings);

    await _localNotifications.initialize(initSettings);

    // 3. Handle Foreground Messages
    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      debugPrint('Received foreground message: ${message.notification?.title}');
      _showLocalNotification(message);
    });

    // 4. Handle Background/Terminated state clicks
    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      debugPrint('Notification clicked: ${message.data}');
      // Handle navigation here if needed
    });
  }

  Future<void> _showLocalNotification(RemoteMessage message) async {
    RemoteNotification? notification = message.notification;
    AndroidNotification? android = message.notification?.android;

    if (notification != null && android != null) {
      await _localNotifications.show(
        notification.hashCode,
        notification.title,
        notification.body,
        const NotificationDetails(
          android: AndroidNotificationDetails(
            'workhub_notifications',
            'WorkHub Notifications',
            importance: Importance.max,
            priority: Priority.high,
          ),
          iOS: DarwinNotificationDetails(),
        ),
        payload: message.data.toString(),
      );
    }
  }

  Future<String?> getToken() async {
    return await _fcm.getToken();
  }

  Future<void> saveTokenToFirestore() async {
    final user = FirebaseAuth.instance.currentUser;
    if (user != null) {
      String? token = await getToken();
      if (token != null) {
        await FirebaseFirestore.instance.collection('users').doc(user.uid).set({
          'fcmToken': token,
          'lastTokenUpdate': FieldValue.serverTimestamp(),
        }, SetOptions(merge: true));
        debugPrint('FCM Token saved to Firestore');
      }
    }
  }

  // Send a notification by writing to Firestore AND calling the Vercel API
  Future<void> sendNotification({
    required String recipientId,
    required String title,
    required String body,
    Map<String, dynamic>? data,
    String? category,
  }) async {
    // 1. Audit/History: Write to Firestore
    try {
      await FirebaseFirestore.instance.collection('notifications').add({
        'recipientId': recipientId,
        'senderId': FirebaseAuth.instance.currentUser?.uid,
        'title': title,
        'body': body,
        'data': data ?? {},
        'category': category,
        'status':
            'pending', // Will stay pending if no trigger, but that's fine for history
        'createdAt': FieldValue.serverTimestamp(),
      });
    } catch (e) {
      debugPrint("Error saving notification to Firestore: $e");
    }

    // 2. Actual Push: Call Vercel API
    try {
      final user = FirebaseAuth.instance.currentUser;
      if (user == null) return;

      final token = await user.getIdToken();
      debugPrint("Sending push notification via Vercel API...");

      final response = await http.post(
        Uri.parse('$_baseUrl/sendPushNotification'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'recipientId': recipientId,
          'title': title,
          'body': body,
          'data': data,
        }),
      );

      if (response.statusCode == 200) {
        debugPrint("Push notification sent successfully.");
      } else {
        debugPrint("Failed to send push notification: ${response.body}");
      }
    } catch (e) {
      debugPrint("Error calling Vercel API: $e");
    }
  }
}
